File: app/src/main/java/bzh/naoned/naonedbusdatadisplay/view/ui/TrafficListFragment.java
package bzh.naoned.naonedbusdatadisplay.view.ui;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import bzh.naoned.naonedbusdatadisplay.MainActivity;
import bzh.naoned.naonedbusdatadisplay.R;
import bzh.naoned.naonedbusdatadisplay.databinding.FragmentTrafficListBinding;
import bzh.naoned.naonedbusdatadisplay.service.model.TrafficData;
import bzh.naoned.naonedbusdatadisplay.view.adapter.TrafficAdapter;
import bzh.naoned.naonedbusdatadisplay.view.callback.TrafficClickCallback;
import bzh.naoned.naonedbusdatadisplay.viewmodel.TrafficListViewModel;

public class TrafficListFragment extends Fragment {
    private bzh.naoned.naonedbusdatadisplay.viewmodel.TrafficListViewModel trafficListViewModel;
    private TrafficAdapter trafficAdapter;
    private FragmentTrafficListBinding binding;
    private TrafficClickCallback trafficClickCallback = new TrafficClickCallback() {
        @Override
        public void onClick(bzh.naoned.naonedbusdatadisplay.service.model.TrafficData trafficData) {
            ((MainActivity) getActivity()).show(trafficData);
        }
    };

    public static TrafficListFragment newInstance() {
        return new TrafficListFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_traffic_list, container, false);

        trafficAdapter = new TrafficAdapter(trafficClickCallback);
        binding.recyclerTraffic.setAdapter(trafficAdapter);
        binding.setIsLoading(true);

        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        trafficListViewModel = ViewModelProviders.of(this).get(bzh.naoned.naonedbusdatadisplay.viewmodel.TrafficListViewModel.class);
        trafficListViewModel.init();

        observeViewModel(trafficListViewModel);
    }

    private void observeViewModel(TrafficListViewModel viewModel) {
        // define observer to update ui when data change
        viewModel.getTrafficListObservable().observe(this, new Observer<List<TrafficData>>() {
            @Override
            public void onChanged(@Nullable List<TrafficData> trafficData) {
                // update ui
                if (trafficData != null) {
                    binding.setIsLoading(false);
                    trafficAdapter.setValues(trafficData);
                }
            }
        });

    }
}

File: app/src/main/java/bzh/naoned/naonedbusdatadisplay/view/ui/TrafficListFragment.kt
package bzh.naoned.naonedbusdatadisplay.view.ui

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import bzh.naoned.naonedbusdatadisplay.MainActivity
import bzh.naoned.naonedbusdatadisplay.R
import bzh.naoned.naonedbusdatadisplay.databinding.FragmentTrafficListBinding
import bzh.naoned.naonedbusdatadisplay.service.model.TrafficData
import bzh.naoned.naonedbusdatadisplay.view.adapter.TrafficAdapter
import bzh.naoned.naonedbusdatadisplay.view.callback.TrafficClickCallback
import bzh.naoned.naonedbusdatadisplay.viewmodel.TrafficListViewModel

class TrafficListFragment : Fragment() {
    private lateinit var trafficListViewModel: bzh.naoned.naonedbusdatadisplay.viewmodel.TrafficListViewModel
    private lateinit var trafficAdapter: TrafficAdapter
    private lateinit var binding: FragmentTrafficListBinding

    private val trafficClickCallback = object: TrafficClickCallback {
        override fun onClick(trafficData: TrafficData) = (activity as MainActivity).show(trafficData)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_traffic_list, container, false)

        trafficAdapter = TrafficAdapter(trafficClickCallback)
        binding.recyclerTraffic.adapter = trafficAdapter
        binding.isLoading = true

        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        trafficListViewModel = ViewModelProviders.of(this).get(bzh.naoned.naonedbusdatadisplay.viewmodel.TrafficListViewModel::class.java)
        trafficListViewModel.init()

        observeViewModel(trafficListViewModel)
    }

    private fun observeViewModel(viewModel: TrafficListViewModel?) {
        // define observer to update ui when data change
        viewModel?.trafficListObservable?.observe(this, Observer { trafficData ->
            // update ui
            if (trafficData != null) {
                binding.isLoading = false
                trafficAdapter.setValues(trafficData)
            }
        })

    }

    companion object {

        fun newInstance(): TrafficListFragment {
            return TrafficListFragment()
        }
    }
}
